sdk/core/core/test/cmocka/test_az_credential_client_secret.c

```diff
@@ -0,0 +1,176 @@
+{
+  az_span const request_url
+      = az_span_slice(request->_internal.url, 0, request->_internal.url_length);
+
+  if (!az_span_is_content_equal(
+          AZ_SPAN_FROM_STR("https://www.microsoft.com/test/request"),
+          request_url)) // Auth request
+  {
+    assert_true(az_span_is_content_equal(
+        AZ_SPAN_FROM_STR("https://login.microsoftonline.com/TenantID/oauth2/v2.0/token"),
+        request_url));
+
+    assert_true(az_span_is_content_equal(
+        AZ_SPAN_FROM_STR("grant_type=client_credentials"
+                         "&client_id=ClientID"
+                         "&scope=Scopes"
+                         "&client_secret=ClientSecret"),
+        request->_internal.body));
+
+    assert_int_equal(1, _az_http_request_headers_count(request));
+    {
+      az_pair header = { 0 };
+
+      az_result const ignore = az_http_request_get_header(request, 0, &header);
+      (void)ignore;
+
+      assert_true(az_span_is_content_equal(AZ_SPAN_FROM_STR("Content-Type"), header.key));
+      assert_true(az_span_is_content_equal(
+          AZ_SPAN_FROM_STR("application/x-www-form-urlencoded"), header.value));
+    }
+
+    static int auth_attempt = 0;
+#ifdef MOCK_ENABLED
+    will_return(__wrap_az_platform_clock_msec, 0);
+    ++auth_attempt;
+#endif // MOCK_ENABLED
+
+    // 3rd attempt should never happen because the token should not be expiring given that clock
+    // returns 0.
+    assert_in_range(auth_attempt, 1, 2);
+
+    if (auth_attempt == 1)
+    {
+      response->_internal.http_response
+          = AZ_SPAN_FROM_STR("HTTP/1.1 500 Internal Server Error\r\n\r\n");
+    }
+    else
+    {
+      response->_internal.http_response = AZ_SPAN_FROM_STR(
+          "HTTP/1.1 200 OK\r\n\r\n{ 'access_token' : 'AccessToken', 'expires_in' : 3600 }");
+    }
+  }
+  else // The actual HTTP request
+  {
+    bool has_auth_header = false;
+    int32_t const header_count = _az_http_request_headers_count(request);
+    for (int32_t i = 0; i < header_count; ++i)
+    {
+      az_pair header = { 0 };
+
+      az_result const ignore = az_http_request_get_header(request, i, &header);
+      (void)ignore;
+
+      if (az_span_is_content_equal(AZ_SPAN_FROM_STR("authorization"), header.key))
+      {
+        assert_true(az_span_is_content_equal(AZ_SPAN_FROM_STR("Bearer AccessToken"), header.value));
+        has_auth_header = true;
+      }
+    }
+
+#ifdef MOCK_ENABLED
+    assert_true(has_auth_header);
+#endif // MOCK_ENABLED
+
+    static bool first_attempt = true;
+    if (first_attempt)
+    {
+      response->_internal.http_response = AZ_SPAN_FROM_STR("FirstResponse");
+      first_attempt = false;
+    }
+    else
+    {
+      response->_internal.http_response = AZ_SPAN_FROM_STR("SubsequentResponse");
+    }
+  }
+
+  return AZ_OK;
+}
+
+#ifdef MOCK_ENABLED
+az_result __wrap_az_http_client_send_request(_az_http_request* request, az_http_response* response)
```

Comment:
Who calls this?

Reply:
CMocka does substitute the next call to az_http_client_send_request() with this. az_http_client_send_request() is called by client secret credential, and by the HTTP pipeline.